// bitski/lib/-private/errors/parse-error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ParseError = exports.ParseErrorCode = void 0;
var ParseErrorCode;
(function (ParseErrorCode) {
    // Could not parse response as JSON
    ParseErrorCode[ParseErrorCode["InvalidJSON"] = 2000] = "InvalidJSON";
    // Received an error response status,
    // but not error body was provided.
    ParseErrorCode[ParseErrorCode["NoErrorBody"] = 2001] = "NoErrorBody";
})(ParseErrorCode = exports.ParseErrorCode || (exports.ParseErrorCode = {}));
/**
 * Represents an error that is thrown during decoding
 */
class ParseError extends Error {
    constructor(message, code) {
        super(message);
        this.name = 'ParseError';
        // Maintains proper stack trace for where our error was thrown (only available on V8)
        if (Error.captureStackTrace) {
            Error.captureStackTrace(this, ParseError);
        }
        this.code = code;
    }
    static InvalidJSON() {
        return new ParseError('Could not decode response as JSON', ParseErrorCode.InvalidJSON);
    }
    static UnknownError() {
        return new ParseError('An unknown error occurred.', ParseErrorCode.NoErrorBody);
    }
}
exports.ParseError = ParseError;
