// bitski/lib/-private/auth/user-store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserStore = void 0;
const constants_1 = require("../constants");
const localstorage_store_1 = require("../utils/localstorage-store");
const user_1 = require("./user");
class UserStore {
    constructor(clientId, store) {
        this.clientId = clientId;
        this.store = store || new localstorage_store_1.LocalStorageStore();
        this.user = this.fetchUser();
    }
    async getCurrentUser() {
        const user = await this.user;
        return user !== null && user !== void 0 ? user : this.fetchUser();
    }
    get storageKey() {
        return `${constants_1.USER_KEY}.${this.clientId}`;
    }
    async set(user) {
        this.user = Promise.resolve(user);
        await this.cacheUser(user);
    }
    async clear() {
        this.user = Promise.resolve(undefined);
        await this.cacheUser(undefined);
    }
    loadUserFromCache() {
        this.user = this.fetchUser();
    }
    async fetchUser() {
        const userData = await this.store.getItem(this.storageKey);
        if (userData) {
            return user_1.User.fromString(userData);
        }
    }
    async cacheUser(user) {
        if (user) {
            await this.store.setItem(this.storageKey, user.toStorageString());
        }
        else {
            await this.store.clearItem(this.storageKey);
        }
    }
}
exports.UserStore = UserStore;
