// @bitski/provider-engine/dist/modules/util/block-tracker.js
import PollingBlockTracker from 'eth-block-tracker';
import { EventEmitter } from 'events';
import { BlockTrackerError } from '../errors/block-tracker-error';
import { createPayload } from './create-payload';
import { toBuffer } from './eth-util';
// Class responsible for tracking new blocks as they are mined,
// loading them, parsing them, and alerting subscribers via events.
export default class BlockTracker extends EventEmitter {
    constructor(opts) {
        super();
        // Number of milliseconds to wait before retrying
        this.blockTimeout = 300;
        // Maximum attempts to load a block
        this.maxBlockRetries = 3;
        this.provider = opts.provider;
        this._blockTracker = opts.blockTracker || new PollingBlockTracker(Object.assign({}, opts, { setSkipCacheFlag: true }));
    }
    start() {
        this.createSubscriptions();
    }
    stop() {
        this.destroySubscriptions();
    }
    getLatestBlock() {
        return this._blockTracker.checkForLatestBlock().catch((error) => {
            this.emit('error', error);
        });
    }
    createSubscriptions() {
        // on new block, request block body and emit as events
        this._blockTracker.on('latest', this.onLatest.bind(this));
        // forward other events
        this._blockTracker.on('sync', this.emit.bind(this, 'sync'));
        this._blockTracker.on('error', this.emit.bind(this, 'error'));
    }
    destroySubscriptions() {
        // stop block polling by removing event listeners
        this._blockTracker.removeAllListeners();
    }
    onLatest(blockNumber) {
        this.currentBlockNumber = blockNumber;
        this.loadBlock(blockNumber);
    }
    // Tries to get the block payload recursively
    loadBlock(blockNumber, callCount = 0) {
        this._getBlockByNumber(blockNumber).then((blockResponse) => {
            // Result can be null if the block hasn't fully propagated to the nodes
            if (blockResponse.result) {
                this.updateBlock(blockResponse.result);
            }
            else if (callCount < this.maxBlockRetries && blockNumber === this.currentBlockNumber) {
                // Only call recursively if the current block number is still the same
                // and if we are under the retry limit.
                setTimeout(() => {
                    this.loadBlock(blockNumber, callCount + 1);
                }, this.blockTimeout);
            }
            else {
                throw BlockTrackerError.BlockNotFound(blockNumber);
            }
        }).catch((err) => {
            // Don't retry for errors (provider should have already retried)
            this.emit('error', err);
        });
    }
    _getBlockByNumber(blockNumber) {
        const req = createPayload({ method: 'eth_getBlockByNumber', params: [blockNumber, false], skipCache: true });
        return new Promise((fulfill, reject) => {
            this.provider.sendAsync(req, (err, result) => {
                if (err) {
                    return reject(err);
                }
                fulfill(result);
            });
        });
    }
    // Parse the block into a buffer representation and update subscribers.
    updateBlock(block) {
        const bufferBlock = toBufferBlock(block);
        // set current + emit "block" event
        this._setCurrentBlock(bufferBlock);
        // emit other events
        this.emit('rawBlock', block);
        this.emit('latest', block);
    }
    _setCurrentBlock(bufferBlock) {
        this.currentBlock = bufferBlock;
        this.emit('block', bufferBlock);
    }
}
// util
function toBufferBlock(jsonBlock) {
    return {
        number: toBuffer(jsonBlock.number),
        hash: toBuffer(jsonBlock.hash),
        parentHash: toBuffer(jsonBlock.parentHash),
        nonce: toBuffer(jsonBlock.nonce),
        mixHash: toBuffer(jsonBlock.mixHash),
        sha3Uncles: toBuffer(jsonBlock.sha3Uncles),
        logsBloom: toBuffer(jsonBlock.logsBloom),
        transactionsRoot: toBuffer(jsonBlock.transactionsRoot),
        stateRoot: toBuffer(jsonBlock.stateRoot),
        receiptsRoot: toBuffer(jsonBlock.receiptsRoot),
        miner: toBuffer(jsonBlock.miner),
        difficulty: toBuffer(jsonBlock.difficulty),
        totalDifficulty: toBuffer(jsonBlock.totalDifficulty),
        size: toBuffer(jsonBlock.size),
        extraData: toBuffer(jsonBlock.extraData),
        gasLimit: toBuffer(jsonBlock.gasLimit),
        gasUsed: toBuffer(jsonBlock.gasUsed),
        timestamp: toBuffer(jsonBlock.timestamp),
        transactions: jsonBlock.transactions,
    };
}
