// web3-react-fortmatic/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fortmatic = void 0;
const types_1 = require("@web3-react/types");
class Fortmatic extends types_1.Connector {
    constructor(actions, { apiKey, network }) {
        super(actions);
        this.apiKey = apiKey;
        this.network = network;
    }
    startListening(configuration) {
        return __awaiter(this, void 0, void 0, function* () {
            return Promise.resolve().then(() => __importStar(require("fortmatic"))).then((m) => __awaiter(this, void 0, void 0, function* () {
                this.fortmatic = new m.default(this.apiKey, this.network);
                const [Web3Provider, Eip1193Bridge] = yield Promise.all([
                    Promise.resolve().then(() => __importStar(require("@ethersproject/providers"))).then(({ Web3Provider }) => Web3Provider),
                    Promise.resolve().then(() => __importStar(require("@ethersproject/experimental"))).then(({ Eip1193Bridge }) => Eip1193Bridge),
                ]);
                const provider = new Web3Provider(this.fortmatic.getProvider());
                this.provider = new Eip1193Bridge(provider.getSigner(), provider);
            }));
        });
    }
    activate(configuration) {
        return __awaiter(this, void 0, void 0, function* () {
            this.actions.startActivation();
            yield this.startListening(configuration).catch((error) => {
                console.error("error", error);
            });
            console.log("this.provider, this.fortmatic", this.provider, this.fortmatic);
            if (this.provider) {
                yield Promise.all([
                    this.provider.request({ method: "eth_chainId" }),
                    this.provider.request({ method: "eth_accounts" }),
                ])
                    .then(([chainId, accounts]) => {
                    this.actions.update({
                        chainId: Number.parseInt(chainId, 16),
                        accounts,
                    });
                    this.fortmatic.user.deposit();
                })
                    .catch((error) => {
                    console.error("error", error);
                });
            }
        });
    }
    /** {@inheritdoc Connector.deactivate} */
    deactivate() {
        return __awaiter(this, void 0, void 0, function* () {
            this.fortmatic.user.logout();
            this.actions.resetState();
        });
    }
}
exports.Fortmatic = Fortmatic;
