// @bitski/provider-engine/dist/modules/base-provider.js
import { EventEmitter } from 'events';
import { ProviderEngineError, ProviderEngineErrorCode } from './errors/provider-engine-error';
import { createPayload } from './util/create-payload';
// The base class which ProviderEngine will extend from that provides the basic Web3 Provider interface
export default class BaseProvider extends EventEmitter {
    // Modern send method
    send(method, params = []) {
        const payload = createPayload({ method, params });
        return this.sendPayload(payload).then((response) => {
            return response.result;
        });
    }
    // Legacy sendAsync method
    sendAsync(payload, cb) {
        this.sendPayload(payload).then((response) => {
            cb(null, response);
        }).catch((error) => {
            cb(error, null);
        });
    }
    // Whether or not this provider supports subscriptions
    supportsSubscriptions() {
        // Override this in your subclass if you support subscriptions
        return false;
    }
    // Method to subscribe to a given subscription type
    subscribe(_subscribeMethod, _subscriptionMethod, _parameters) {
        // Override this with subscription implementation
        return Promise.reject(new ProviderEngineError('Subscriptions are not supported', ProviderEngineErrorCode.UnsupportedFeature));
    }
    // Method to unsubscribe
    unsubscribe(_subscriptionId, _unsubscribeMethod) {
        // Override this with unsubscribe implementation
        return Promise.reject(new ProviderEngineError('Subscriptions are not supported', ProviderEngineErrorCode.UnsupportedFeature));
    }
}
