// bitski/dist/esm/-private/auth/token-store.js
import { ACCESS_TOKEN_KEY, ID_TOKEN_KEY, REFRESH_TOKEN_KEY } from '../constants';
import { LocalStorageStore } from '../utils/localstorage-store';
import { AccessToken } from './access-token';
export class TokenStore {
    constructor(clientId, store) {
        var _a, _b;
        this.clientId = clientId;
        this.store = store || new LocalStorageStore();
        this.loadTokensFromCache();
        (_b = (_a = this.store).onUpdate) === null || _b === void 0 ? void 0 : _b.call(_a, () => this.loadTokensFromCache());
    }
    async getCurrentToken() {
        const accessToken = await this.accessToken;
        if (accessToken && !accessToken.expired) {
            return accessToken.token;
        }
    }
    async getCurrentIdToken() {
        const idToken = await this.idToken;
        const accessToken = await this.accessToken;
        if (idToken && accessToken && !accessToken.expired) {
            return idToken;
        }
    }
    getRefreshToken() {
        return this.refreshToken;
    }
    get idTokenKey() {
        return `${ID_TOKEN_KEY}.${this.clientId}`;
    }
    get accessTokenKey() {
        return `${ACCESS_TOKEN_KEY}.${this.clientId}`;
    }
    get refreshTokenKey() {
        return `${REFRESH_TOKEN_KEY}.${this.clientId}`;
    }
    loadTokensFromCache() {
        this.accessToken = this.store.getItem(this.accessTokenKey).then((accessTokenString) => {
            if (accessTokenString) {
                return AccessToken.fromString(accessTokenString);
            }
        });
        this.idToken = this.store.getItem(this.idTokenKey);
        this.refreshToken = this.store.getItem(this.refreshTokenKey);
    }
    persistTokenResponse(response) {
        const parsedToken = AccessToken.fromTokenResponse(response);
        this.store.setItem(this.accessTokenKey, parsedToken.toStorageString());
        this.accessToken = Promise.resolve(parsedToken);
        this.store.setItem(this.idTokenKey, response.idToken);
        this.idToken = Promise.resolve(response.idToken);
        if (response.refreshToken) {
            this.store.setItem(this.refreshTokenKey, response.refreshToken);
            this.refreshToken = Promise.resolve(response.refreshToken);
        }
    }
    async invalidateCurrentToken() {
        this.accessToken = Promise.resolve(undefined);
        this.idToken = Promise.resolve(undefined);
        await Promise.all([
            this.store.clearItem(this.accessTokenKey),
            this.store.clearItem(this.idTokenKey),
        ]);
    }
    async clear() {
        this.accessToken = Promise.resolve(undefined);
        this.idToken = Promise.resolve(undefined);
        this.refreshToken = Promise.resolve(undefined);
        await Promise.all([
            this.store.clearItem(this.refreshTokenKey),
            this.store.clearItem(this.accessTokenKey),
            this.store.clearItem(this.idTokenKey),
        ]);
    }
}
