// bitski/lib/-private/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthenticationStatus = exports.OAuthSignInMethod = exports.DEFAULT_AUTHORIZED_METHODS = exports.CACHED_METHODS = exports.USER_KEY = exports.ID_TOKEN_KEY = exports.ACCESS_TOKEN_KEY = exports.REFRESH_TOKEN_KEY = exports.DEFAULT_POPUP_FEATURES = exports.CHECK_FOR_POPUP_CLOSE_INTERVAL = exports.DEFAULT_OPTIONAL_SCOPES = exports.DEFAULT_SCOPES = exports.DEFAULT_OAUTH_CONFIGURATION = exports.IFRAME_MESSAGE_ORIGIN_INCLUDES = exports.BITSKI_WEB_BASE_URL = exports.BITSKI_RPC_BASE_URL = exports.BITSKI_TRANSACTION_API_BASE_URL = exports.BITSKI_USER_API_HOST = exports.SDK_VERSION = void 0;
// SDK
exports.SDK_VERSION = '0.14.1';
// URLs
exports.BITSKI_USER_API_HOST = 'https://www.bitski.com/v1';
exports.BITSKI_TRANSACTION_API_BASE_URL = 'https://api.bitski.com/v1';
exports.BITSKI_RPC_BASE_URL = 'https://api.bitski.com/v1/web3';
exports.BITSKI_WEB_BASE_URL = 'https://sign.bitski.com';
exports.IFRAME_MESSAGE_ORIGIN_INCLUDES = '.bitski.com';
// OAuth
exports.DEFAULT_OAUTH_CONFIGURATION = {
    authorization_endpoint: 'https://account.bitski.com/oauth2/auth',
    revocation_endpoint: '',
    token_endpoint: 'https://account.bitski.com/oauth2/token',
    userinfo_endpoint: 'https://account.bitski.com/userinfo',
};
exports.DEFAULT_SCOPES = ['openid']; // scopes that are always included
exports.DEFAULT_OPTIONAL_SCOPES = ['offline']; // scopes that are included by default, but can be overridden
// Popup Window
exports.CHECK_FOR_POPUP_CLOSE_INTERVAL = 500;
exports.DEFAULT_POPUP_FEATURES = {
    location: 'no',
    toolbar: 'no',
    width: 500,
    height: 500,
    left: 100,
    top: 100,
};
// Storage
exports.REFRESH_TOKEN_KEY = 'bitski.refresh_token';
exports.ACCESS_TOKEN_KEY = 'bitski.access_token';
exports.ID_TOKEN_KEY = 'bitski.id_token';
exports.USER_KEY = 'bitski.user';
// Methods
exports.CACHED_METHODS = ['eth_accounts'];
exports.DEFAULT_AUTHORIZED_METHODS = [
    'eth_sendTransaction',
    'eth_signTransaction',
    'eth_sign',
    'personal_sign',
    'eth_signTypedData',
    'eth_signTypedData_v3',
    'eth_signTypedData_v4',
];
var OAuthSignInMethod;
(function (OAuthSignInMethod) {
    OAuthSignInMethod["Redirect"] = "REDIRECT";
    OAuthSignInMethod["Popup"] = "POPUP";
    OAuthSignInMethod["Silent"] = "SILENT";
})(OAuthSignInMethod = exports.OAuthSignInMethod || (exports.OAuthSignInMethod = {}));
var AuthenticationStatus;
(function (AuthenticationStatus) {
    AuthenticationStatus["Connected"] = "CONNECTED";
    AuthenticationStatus["Expired"] = "EXPIRED";
    AuthenticationStatus["NotConnected"] = "NOT_CONNECTED";
})(AuthenticationStatus = exports.AuthenticationStatus || (exports.AuthenticationStatus = {}));
